https: Use secureProtocol in Agent#getName

papandreou opened the pull request:
##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j8 test` (UNIX), or `vcbuild test nosign` (Windows) passes
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)

https

##### Description of change

Fix oversight pointed out by @bnoordhuis in #9324. The value of the `secureProtocol` option should be used as a "discriminator" when determining whether an existing HTTPS connection can be reused.

bnoordhuis commented on test/parallel/test-https-agent-secure-protocol.js:
```diff
@@ -0,0 +1,67 @@
+'use strict';
+var assert = require('assert');
+var common = require('../common');
```
Can you use `const` here and elsewhere where it makes sense?

bnoordhuis commented on test/parallel/test-https-agent-secure-protocol.js:
```diff
@@ -0,0 +1,67 @@
+    secureProtocol: 'SSLv23_method'
+  }, function(res) {
+    res.resume();
+    globalAgent.once('free', function () {
```
Style issue: can you drop the blank in `function ()`.  Same issue a few lines below.

bnoordhuis commented on test/parallel/test-https-agent-secure-protocol.js:
```diff
@@ -0,0 +1,67 @@
+          var keys = Object.keys(globalAgent.freeSockets);
+          assert.equal(keys.length, 2);
+          assert.notEqual(keys[0].indexOf(':SSLv23_method'), -1);
+          assert.notEqual(keys[1].indexOf(':TLSv1_method'), -1);
```
Can you use `assert.strictEqual` and `assert.notStrictEqual` here?  Also, instead of .indexOf() you can use .includes().

bnoordhuis replied:
This is still unfixed.

bnoordhuis commented on test/parallel/test-https-agent-secure-protocol.js:
```diff
@@ -0,0 +1,67 @@
+        });
+      }).on('error', function(e) {
+        console.log(e.message);
+        process.exit(1);
```
Don't do this, just let the exception bubble up.

papandreou commented:
@bnoordhuis, thanks for taking a look! I believe I fixed the problems you pointed out now.

bnoordhuis commented on test/parallel/test-https-agent-secure-protocol.js:
```diff
@@ -0,0 +1,61 @@
+var options = {
+  key: fs.readFileSync(common.fixturesDir + '/keys/agent1-key.pem'),
+  cert: fs.readFileSync(common.fixturesDir + '/keys/agent1-cert.pem'),
+  ca:  fs.readFileSync(common.fixturesDir + '/keys/ca1-cert.pem')
```
Tiny nit: there is an extra blank here.

bnoordhuis commented on test/parallel/test-https-agent-secure-protocol.js:
```diff
@@ -0,0 +1,61 @@
+      });
+    });
+  });
+});
```
There are still a couple of places where you could s/var/const/.

papandreou commented:
@bnoordhuis ... There, that should do it.

For some reason I didn't notice that the test-https-agent-getname.js was also affected by this (due to the extra colons showing up in those keys). Fixed that as well and squashed+rebased.

cjihrig commented:
One nit: you might use `common.mustCall()` to ensure that the callbacks are actually being invoked and your assertions are running.

papandreou commented:
@cjihrig Good point -- I added `common.mustCall` and force pushed now.

jasnell commented:
Looks like this fell through the cracks. New CI: https://ci.nodejs.org/job/node-test-pull-request/5804/

jasnell commented:
Landed in a469f85

MylesBorins commented:
Is this something we want to backport? It seems like it changes some behavior... even if it does so to a more correct behavior

jasnell commented:
hmmm.. I'd say maybe let it sit for a while longer before putting into v6.x-staging. Perhaps skip v4.x-staging

MylesBorins commented:
It seems like this PR has baked long enough. Landing on v6.x
